lib/internal/http2.js

```diff
@@ -48,7 +46,139 @@ Object.defineProperty(exports, 'constants', {
 // emit events as callbacks. To allow this to happen,
 // it has to inherit from EventEmitter.
 util.inherits(http2.Http2Session, EventEmitter);
+util.inherits(http2.Http2Stream, Duplex);
 
+function maybeDestroyStream(stream) {
+  if ((!stream.readable &&
+       !stream.writable &&
+       !stream._writableState.length)) {
+    stream.destroy();
+  }
+}
+
+http2.Http2Stream.prototype.onread = function onread(nread, buffer) {
+  const stream = this;
+
+  if (nread > 0) {
+    var ret = stream.push(buffer);
+    if (stream.reading && !ret) {
+      stream.reading = false;
+      var err = stream.readStop();
+      if (err) {
+        // TODO(jasnell): figure this out
+        maybeDestroyStream(stream);
+      }
+    }
+    return;
+  }
+
+  if (nread === 0) {
+    return;
+  }
+
+  if (nread !== uv.UV_EOF) {
+    // TODO(jasnell): figure out
+    return maybeDestroyStream(stream);
+  }
+
+  stream.push(null);
+
+  if (stream._readableState.length === 0) {
+    stream.readable = false;
+    // TODO(jasnell): Figure out
+    //maybeDestroy(self);
+  }
+};
+
+http2.Http2Stream.prototype.end = function(chunk, encoding, callback) {
+  Duplex.prototype.end.call(this, chunk, encoding, callback);
+  var req = new ShutdownWrap();
+  req.oncomplete = (status, handle, req) => {
+    process.nextTick(() => maybeDestroyStream(this));
+  };
+  req.handle = this;
+  this.shutdown(req);
+};
+
+http2.Http2Stream.prototype._read = function(size) {
+  this.reading = true;
+  this.readStart();
+};
+
+http2.Http2Stream.prototype._writeGeneric = function(writev,
+                                                     data,
+                                                     encoding,
+                                                     cb) {
+  const req = new WriteWrap();
+  req.handle = this;
+  req.oncomplete = () => {};
+  req.async = false;
+  var err;
+
+  if (writev) {
+    const chunks = new Array(data.length << 1);
+    for (var i = 0; i < data.length; i++) {
+      const entry = data[i];
+      chunks[i * 2] = entry.chunk;
+      chunks[i * 2 + 1] = entry.encoding;
+    }
+    err = this.writev(req, chunks);
+
+    // Retain chunks
+    if (err === 0) req._chunks = chunks;
+  } else {
+    var enc;
+    if (data instanceof Buffer) {
+      enc = 'buffer';
+    } else {
+      enc = encoding;
+    }
+    err = createWriteReq(req, this, data, enc);
+  }
+
+  if (err) {
+    throw util._errnoException(err, 'write', req.error);
+  }
+
+  this._bytesDispatched += req.bytes;
+
+  cb();
```

Comment:
should this `cb` be called in `req.oncomplete`? Otherwise we will not be handling backpressure?